Hw10 asm

aterDex opened the pull request.

Commit:
hw10 start

Commit:
Добавление лога в консоль, без имени параметров

Commit:
изменил правила определения статического поля

Commit:
Добавил поддержку наименование параметров

Commit:
Удалил лишнии импорты

Commit:
Сделала номарльного агента, создаль модуль для демонстрации работы агента. Настроил зависимости между модулями.

Commit:
Исправил NullPointerException при обработке аргументов агента.
Подправил тест

Commit:
Коментарий +

Commit:
Уплотнили тесты.

Commit:
Обработка параметров у task run, для запуска приложение с определением имен параметров, или без определения имен параметров.

Commit:
Чуть изменили строчку формирования лога

Commit:
Вынесли небольшой блок в новый метод, для лучшей читабельности

Commit:
Вызов некоторых методов, результат которых никогда не меняется, перенс в константы

Commit:
Opcodes сделал static import для краткости

Commit:
Еще чучуть уплотнил тесты

Commit:
Научил градл запускать приложение без агента

Commit:
Добавил README для hw10-asm-app

Commit:
Правка README для hw10-asm-app

Commit:
Добавил README для hw10-asm

Commit:
Отрефакторил тесты

Commit:
Перенес зависимость от друго модуля по флаг

Commit:
Переработал работу с параметрами объектами

Commit:
Вынес логику по вставке лога в отдельный класс

Commit:
Написал тесты для HeraldInjectorParametersLogBySystemOut

Commit:
убрал лишнии коментарии

Commit:
Сократил метод проверки объекта

Commit:
Добавили свойства для аннотации Log

Commit:
Тест для проверки printProperty

Commit:
Тесты для comment

Commit:
Add Comment in log

petrelevich commented on hw10-asm/src/main/java/ru/otus/homework/herald/core/HeraldClassTransformer.java:
```diff
@@ -0,0 +1,76 @@
+ * В этом режиме мы проходим по классу только раз, и если атрибут есть у метода
+ * то вставляем в начало метода вывод параметров, при этом имеена параметров будут parN где N порядковый номер параметра.
+ * <p>
+ * В режиме resolveParameterName = true, мы проходим по классу дважды,
```
по классу надо ходить один раз и за один проход формировать нормальную строку лога.
Т.е. должен быть один режим, совмещающий ваши два.

aterDex replied:
Сделал реализацию в один проход в классе HeraldClassVisitorLogMethod
Преведущий класс HeraldClassVisitor переименовал в HeraldClassVisitorInline

petrelevich commented on hw10-asm/src/main/java/ru/otus/homework/herald/core/HeraldClassVisitor.java:
```diff
@@ -0,0 +1,48 @@
+    private final Collection<HeraldMeta> heralds;
+    private final Supplier<Optional<HeraldInjector>> injectorForMethodFactory;
+
+    public HeraldClassVisitor(int api, ClassVisitor classVisitor, Collection<HeraldMeta> heralds, Supplier<Optional<HeraldInjector>> injectorForMethodFactory) {
```
зачем тут supplier?
почему прямо в конструкторе нельзя создать
new HeraldInjectorParametersLogBySystemOut()
и зачем Optional?

aterDex replied:
Перемудрил я)
Думал использовать как фабрику, Optional - чтобы явно сказать что null там быть не может.
Убрал этот supplier, создаю HeraldInjectorParametersLogBySystemOut на месте, но не в конструкторе, т.к. не для каждого метода он нужен, а только для тех у кого аннтоция найденна. Класс сейчас называется HeraldClassVisitorInline

petrelevich commented on hw10-asm/src/main/java/ru/otus/homework/herald/core/HeraldClassVisitor.java:
```diff
@@ -0,0 +1,48 @@
+
+    public HeraldClassVisitor(int api, ClassVisitor classVisitor, Collection<HeraldMeta> heralds, Supplier<Optional<HeraldInjector>> injectorForMethodFactory) {
+        super(api, classVisitor);
+        assert Objects.nonNull(injectorForMethodFactory);
```
обычно в production проверку assert отключают

aterDex replied:
Убрал их.

petrelevich commented on hw10-asm/src/main/java/ru/otus/homework/herald/core/HeraldInjectorParametersLogBySystemOut.java:
```diff
@@ -0,0 +1,149 @@
+public class HeraldInjectorParametersLogBySystemOut implements HeraldInjector {
+
+    public static final Type PRINT_STREAM = Type.getType(PrintStream.class);
+    public static final Type STRING_BUILDER = Type.getType(StringBuilder.class);
```
У вас не должно быть StringBuilder
используйте java.lang.invoke.StringConcatFactory как в примере к вебинару.

aterDex replied:
Реализовал данный подход в классе HeraldInjectorParametersLogBySystemOutOverStringConcatFactory

petrelevich commented on hw10-asm/src/main/java/ru/otus/homework/herald/core/HeraldInjectorParametersLogBySystemOut.java:
```diff
@@ -0,0 +1,149 @@
+     */
+    private String resolveName(int idx, int idxOnStack, HeraldMeta heraldMeta) {
+        if (heraldMeta != null && !heraldMeta.getLocalVariables().isEmpty()) {
+            Optional<LocalVariableMeta> localVar = heraldMeta.getLocalVariables().parallelStream()
```
parallelStream в таком низкоуровневом коде лучше не использовать.

aterDex replied:
Убрал parallelStream

petrelevich commented on hw10-asm/src/main/java/ru/otus/homework/herald/core/HeraldInjectorParametersLogBySystemOut.java:
```diff
@@ -0,0 +1,149 @@
+            Optional<LocalVariableMeta> localVar = heraldMeta.getLocalVariables().parallelStream()
+                    .filter(x -> x.getIndex() == idxOnStack)
+                    .findAny();
+            if (localVar.isPresent()) {
```
Сразу после Optional 
if (localVar.isPresent()) {
выглядит странно.

aterDex replied:
Не хотел дважды использовать конструкцию **"par" + idx**
Переписал на такое
```
if (heraldMeta != null) {
            return heraldMeta.getLocalVariables().stream()
                    .filter(x -> x.getIndex() == idxOnStack)
                    .findAny()
                    .map(x -> x.getName())
                    .orElse("par" + idx);
        }
        return "par" + idx;
```

petrelevich commented on hw10-asm/src/main/java/ru/otus/homework/herald/core/HeraldDetectorMethodVisitor.java:
```diff
@@ -0,0 +1,74 @@
+                        break;
+                    case "comment":
+                        comment = value.toString();
+                        break;
```
хорошо бы секцию break добавить.

aterDex replied:
Я так понимаю вы имели в виду секцию default? Придумал там только бросать исключение)
```
default:
                        throw new UnsupportedOperationException("Переменная " + name + " не поддерживается для аннотации " + LOG_ANNOTATION);
```

petrelevich commented on hw10-asm/src/main/java/ru/otus/homework/herald/core/HeraldDetectorMethodVisitor.java:
```diff
@@ -0,0 +1,74 @@
+        if (!codeHaveBeenVisited) {
+            Type t = Type.getType(descriptor);
+            herald = herald || LOG_ANNOTATION.equals(t.getClassName());
+            return new HeraldLogAnnotationVisitor(api, defaultVisitor);
```
я правильно понимаю, что для каждой аннотации в jdk создастся эксемпляр этого класса?

aterDex replied:
Да действительно была такая проблема. Исправил.

petrelevich commented on hw10-asm/src/main/java/ru/otus/homework/herald/core/HeraldMethodVisitor.java:
```diff
@@ -0,0 +1,48 @@
+        super.visitCode();
+        try {
+            if (!visitCodeBarrier && (heraldMeta != null || isHerald())) {
+                visitCodeBarrier = true;
```
не понятно зачем visitCodeBarrier = true;
если в finally  будет  visitCodeBarrier = false;

aterDex replied:
В метод inject мы передаем visitor в котором сами и находимся (this). Если вдруг кто то в нутри метода inject вызовет метод visitCode(), то произойдет закольцовка. Чтобы ее не было мы перед вызовом inject выставляем флаг, а после уже ее работы его убираем.
Хотя конечно можно и не возращать флаг обратно в false, всеравно для метода этот inject должен произойти лишь раз и последующие вызовы visitCode если их кто то дополнительно вызовет не должны приводить к повтороной вставке. Не уверне как тут лучше поступить.

petrelevich commented on hw10-asm/src/main/java/ru/otus/homework/herald/core/HeraldClassVisitor.java:
```diff
@@ -0,0 +1,48 @@
+ */
+public class HeraldClassVisitor extends ClassVisitor {
+
+    private final Collection<HeraldMeta> heralds;
```
не понятное название переменной,
хорошо бы переименовать более говоряще.

aterDex replied:
Переименовал.

Commit:
Убрал Supplier<Optional<HeraldInjector>>injectorForMethodFactory
https://github.com/aterDex/homework-java/pull/4#discussion_r465372932

Commit:
parallelStream заменил на stream
https://github.com/aterDex/homework-java/pull/4#discussion_r465376427

Commit:
Переписал резолв мета информации
https://github.com/aterDex/homework-java/pull/4#discussion_r465376771

Commit:
Включаем обработчик аннотации только для нужного класса
https://github.com/aterDex/homework-java/pull/4#discussion_r465378080

Commit:
Переменовал переменную
https://github.com/aterDex/homework-java/pull/4#discussion_r465380477

Commit:
Добавили реализацию HeraldClassVisitorLogMethod, который создает строчку лога в один проход.
HeraldClassVisitor переимановали в HeraldClassVisitorInline, его реализация осталась по больше части тойже. Как альтернативный метод к тойже задаче.
Uber jar создается с реализацией HeraldClassVisitorLogMethod
https://github.com/aterDex/homework-java/pull/4#discussion_r465370419

Commit:
Добавил секцию default при обработки аннотации лог

https://github.com/aterDex/homework-java/pull/4#discussion_r465377331

Commit:
Добавил секцию default при обработки аннотации лог +

https://github.com/aterDex/homework-java/pull/4#discussion_r465377331

Commit:
Добавил обработку visitParameter при поиске методов для логирования. Реализовал в HeraldClassVisitorLogMethod, две опции обработки, если visitParameter были, и если их небыло.

Commit:
Вынес дублирущийся код в класс HeraldUtils

petrelevich commented on hw10-asm/src/main/java/ru/otus/homework/herald/core/SearcherHeraldClassVisitor.java:
```diff
@@ -0,0 +1,42 @@
+public class SearcherHeraldClassVisitor extends ClassVisitor {
+
+    @Getter
+    private List<HeraldMeta> heralds = new ArrayList<>();
```
final можно добавить.

aterDex replied:
Сделал.

petrelevich commented on hw10-asm/src/main/java/ru/otus/homework/herald/core/HeraldDetectorMethodVisitor.java:
```diff
@@ -0,0 +1,80 @@
+import org.objectweb.asm.Type;
+import ru.otus.homework.herald.api.Log;
+
+public class HeraldDetectorMethodVisitor extends MethodVisitor {
```
не очевиден смысл наследования.
Кажется, что это способ избежать копипаста.

P.S.
Зачем в имени каждого класса и большинства переменных повторять Herald ?

aterDex replied:
Возможно. В целом из кода который получился, видно что можно переработать чтобы вышло более компактно.

Вообще да, когда много таких повторений выглядит действительно не очень. Но помогает быстрее использовать подсказки студии при написании кода, тоесть ты пишешь Herald, и видишь все классы которые ты написал...
Ну а именна переменных вытекают уже из названия классов...
Вобщем так лучше не делать, читабильность падает)

petrelevich commented on hw10-asm/src/main/java/ru/otus/homework/herald/core/HeraldInjectorParametersLogBySystemOutOverStringConcatFactory.java:
```diff
@@ -0,0 +1,61 @@
+    }
+
+    private String createPattern(Type[] types, HeraldMeta heraldMeta, String prefix, String comment) {
+        StringBuilder sb = new StringBuilder();
```
применение тут  StringBuilder несколько перечеркивает смысл использования StringConcatFactory выше.

aterDex replied:
Ну не точтобы, всетаки этот StringBuilder, пройдет лишь при старте один раз, чтобы сформировать шаблон для StringConcatFactory.

Я переписал на stream и join. Получилось короче, но не уверен что красиво.

petrelevich commented on hw10-asm/src/main/java/ru/otus/homework/herald/core/LocalVariableMeta.java:
```diff
@@ -0,0 +1,14 @@
+
+@Value
+@Builder(toBuilder = true)
+public class LocalVariableMeta {
```
вы ведь с параметрами методов работаете, почему классс называется LocalVariable?

aterDex replied:
Потому что изначально я работал именно c LocalVariable, и собственно и продолжаю его использовать при обработке как LocalVariable так и Parameter. При этом структура класса соответствует данным именно из LocalVariable
(В Parameter нет индекса переменной и типа переменной, их достраиваю из descriptor). Так что в действительности я всегда возвращаю данные как в секции LocalVariable.

petrelevich commented on hw10-asm/src/main/java/ru/otus/homework/herald/core/HeraldUtils.java:
```diff
@@ -0,0 +1,40 @@
+     */
+    public static String resolveName(int idx, int idxOnStack, HeraldMeta heraldMeta) {
+        if (heraldMeta != null) {
+            return heraldMeta.getLocalVariables().stream()
```
Возможо, для задачи поиска по индексу Map лучше подошла бы.

aterDex replied:
Как вы и рассказывали на лекции про принципы построения Map, здесь бы лучше подошла не просто Map, а самодельная Map. Этот индекс является идеальной хэш функцией, нам четко известно что он никогда не повторится в рамках работы с одним методом! Более того он линейный, начинающийся всегда с 0. Единственно что после long и double идет скачок на +1 элемент. Ну и так как у нас всегда в итоге фиксированный набор элементов количество которых мы можем получить заранее из descriptor. Мы в целом можем тут использовать просто массив, выравненный по индексам (после long и double, будет всегда пустой элемент массива), ну и накрутить на него Map как интерфейс для работы с ним.
Или мы можем оставить все как есть и поработать с поиском, для элемента с индексом N мы можем точно сказать что его значения не может лежать в колеекции раньше чем по индексу  масива M (где N=M) и позже чем M*2, и искать нужно только в этом диапазоне.

Commit:
final +add
https://github.com/aterDex/homework-java/pull/4#discussion_r466018740

Commit:
Изменил StringBuilder на join в stream
https://github.com/aterDex/homework-java/pull/4#discussion_r466025017

aterDex committed:
Merge branch 'master' into hw10-asm
